File: backend-java/src/main/java/com/peopleminer/github/RateLimiterService.java
package com.peopleminer.github;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.Map;
import java.util.concurrent.TimeUnit;

@Service
@Slf4j
public class RateLimiterService {

    private static final String RATE_LIMIT_KEY = "github:rate_limit";
    private static final int BUFFER_SECONDS = 2;
    private static final int MIN_REMAINING_THRESHOLD = 100;

    private final RedisTemplate<String, Object> redisTemplate;

    public RateLimiterService(RedisTemplate<String, Object> redisTemplate) {
        this.redisTemplate = redisTemplate;
    }

    public RateLimitState getState() {
        Map<Object, Object> data = redisTemplate.opsForHash().entries(RATE_LIMIT_KEY);

        if (data == null || data.isEmpty()) {
            return RateLimitState.builder()
                .remaining(5000)
                .limit(5000)
                .resetAt(System.currentTimeMillis() / 1000 + 3600)
                .build();
        }

        return RateLimitState.builder()
            .remaining(parseIntSafe(data.get("remaining"), 5000))
            .limit(parseIntSafe(data.get("limit"), 5000))
            .resetAt(parseLongSafe(data.get("resetAt"), System.currentTimeMillis() / 1000 + 3600))
            .build();
    }

    private int parseIntSafe(Object value, int defaultValue) {
        if (value == null) return defaultValue;
        try {
            if (value instanceof Number) return ((Number) value).intValue();
            return Integer.parseInt(value.toString());
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    private long parseLongSafe(Object value, long defaultValue) {
        if (value == null) return defaultValue;
        try {
            if (value instanceof Number) return ((Number) value).longValue();
            return Long.parseLong(value.toString());
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    public void updateFromHeaders(Map<String, String> headers) {
        String remaining = headers.get("x-ratelimit-remaining");
        String limit = headers.get("x-ratelimit-limit");
        String resetAt = headers.get("x-ratelimit-reset");

        if (remaining != null) {
            redisTemplate.opsForHash().put(RATE_LIMIT_KEY, "remaining", remaining);
            int rem = Integer.parseInt(remaining);
            if (rem <= MIN_REMAINING_THRESHOLD && rem > 0) {
                log.warn("Rate limit low: {} remaining", rem);
            } else if (rem == 0) {
                log.warn("Rate limit exhausted. Resets at {}", resetAt);
            }
        }
        if (limit != null) {
            redisTemplate.opsForHash().put(RATE_LIMIT_KEY, "limit", limit);
        }
        if (resetAt != null) {
            redisTemplate.opsForHash().put(RATE_LIMIT_KEY, "resetAt", resetAt);
        }
    }

    public RateLimitCheck canMakeRequest() {
        RateLimitState state = getState();
        long now = System.currentTimeMillis() / 1000;

        if (now >= state.getResetAt()) {
            redisTemplate.opsForHash().put(RATE_LIMIT_KEY, "remaining", String.valueOf(state.getLimit()));
            redisTemplate.opsForHash().put(RATE_LIMIT_KEY, "resetAt", String.valueOf(now + 3600));
            return RateLimitCheck.builder().allowed(true).build();
        }

        if (state.getRemaining() <= 0) {
            long waitMs = (state.getResetAt() - now + BUFFER_SECONDS) * 1000;
            return RateLimitCheck.builder().allowed(false).waitMs(waitMs).build();
        }

        return RateLimitCheck.builder().allowed(true).build();
    }

    public void decrementRemaining() {
        redisTemplate.opsForHash().increment(RATE_LIMIT_KEY, "remaining", -1);
    }

    public int getRemainingRequests() {
        return getState().getRemaining();
    }

    public RateLimitStatusResponse getStatus() {
        RateLimitState state = getState();
        long now = System.currentTimeMillis() / 1000;

        return RateLimitStatusResponse.builder()
            .remaining(state.getRemaining())
            .limit(state.getLimit())
            .resetAt(LocalDateTime.ofEpochSecond(state.getResetAt(), 0, ZoneOffset.UTC))
            .secondsUntilReset(Math.max(0, state.getResetAt() - now))
            .build();
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class RateLimitState {
        private int remaining;
        private int limit;
        private long resetAt;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class RateLimitCheck {
        private boolean allowed;
        private Long waitMs;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class RateLimitStatusResponse {
        private int remaining;
        private int limit;
        private LocalDateTime resetAt;
        private long secondsUntilReset;
    }
}

File: backend/src/github/rate-limiter.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import Redis from 'ioredis';
import { RateLimitError } from '../common/circuit-breaker.service';

interface RateLimitState {
  remaining: number;
  limit: number;
  resetAt: number;
}

const RATE_LIMIT_KEY = 'github:rate_limit';
const BUFFER_SECONDS = 2;
const MIN_REMAINING_THRESHOLD = 100;

@Injectable()
export class RateLimiterService {
  private readonly logger = new Logger(RateLimiterService.name);
  private redis: Redis;

  constructor(private configService: ConfigService) {
    this.redis = new Redis({
      host: this.configService.get('REDIS_HOST', 'localhost'),
      port: this.configService.get('REDIS_PORT', 6380),
    });
  }

  async getState(): Promise<RateLimitState> {
    const data = await this.redis.hgetall(RATE_LIMIT_KEY);

    if (!data || Object.keys(data).length === 0) {
      return {
        remaining: 5000,
        limit: 5000,
        resetAt: Math.floor(Date.now() / 1000) + 3600,
      };
    }

    return {
      remaining: parseInt(data.remaining || '5000'),
      limit: parseInt(data.limit || '5000'),
      resetAt: parseInt(data.resetAt || String(Math.floor(Date.now() / 1000) + 3600)),
    };
  }

  async updateFromHeaders(headers: {
    'x-ratelimit-remaining'?: string;
    'x-ratelimit-limit'?: string;
    'x-ratelimit-reset'?: string;
  }): Promise<void> {
    const remaining = headers['x-ratelimit-remaining'];
    const limit = headers['x-ratelimit-limit'];
    const resetAt = headers['x-ratelimit-reset'];

    const updates: Record<string, string> = {};

    if (remaining !== undefined) updates.remaining = remaining;
    if (limit !== undefined) updates.limit = limit;
    if (resetAt !== undefined) updates.resetAt = resetAt;

    if (Object.keys(updates).length > 0) {
      await this.redis.hset(RATE_LIMIT_KEY, updates);
    }

    if (remaining !== undefined) {
      const rem = parseInt(remaining);
      if (rem <= MIN_REMAINING_THRESHOLD && rem > 0) {
        this.logger.warn(`Rate limit low: ${rem} remaining`);
      } else if (rem === 0) {
        this.logger.warn(`Rate limit exhausted. Resets at ${new Date(parseInt(resetAt || '0') * 1000).toISOString()}`);
      }
    }
  }

  async waitForRateLimit(): Promise<void> {
    const state = await this.getState();
    const now = Math.floor(Date.now() / 1000);

    if (now >= state.resetAt) {
      await this.redis.hset(RATE_LIMIT_KEY, {
        remaining: String(state.limit),
        resetAt: String(now + 3600),
      });
      return;
    }

    if (state.remaining <= 0) {
      const resetAtMs = state.resetAt * 1000;
      throw new RateLimitError(
        `GitHub rate limit exhausted. Resets at ${new Date(resetAtMs).toISOString()}`,
        resetAtMs + BUFFER_SECONDS * 1000,
      );
    }
  }

  async canMakeRequest(): Promise<{ allowed: boolean; waitMs?: number }> {
    const state = await this.getState();
    const now = Math.floor(Date.now() / 1000);

    if (now >= state.resetAt) {
      await this.redis.hset(RATE_LIMIT_KEY, {
        remaining: String(state.limit),
        resetAt: String(now + 3600),
      });
      return { allowed: true };
    }

    if (state.remaining <= 0) {
      const waitMs = (state.resetAt - now + BUFFER_SECONDS) * 1000;
      return { allowed: false, waitMs };
    }

    return { allowed: true };
  }

  async decrementRemaining(): Promise<void> {
    await this.redis.hincrby(RATE_LIMIT_KEY, 'remaining', -1);
  }

  async getRemainingRequests(): Promise<number> {
    const state = await this.getState();
    return state.remaining;
  }

  async getStatus(): Promise<{ remaining: number; limit: number; resetAt: Date; secondsUntilReset: number }> {
    const state = await this.getState();
    const now = Math.floor(Date.now() / 1000);
    return {
      remaining: state.remaining,
      limit: state.limit,
      resetAt: new Date(state.resetAt * 1000),
      secondsUntilReset: Math.max(0, state.resetAt - now),
    };
  }

  async onModuleDestroy(): Promise<void> {
    await this.redis.quit();
  }
}
